src/chainstate/stacks/index/marf.rs

```diff
@@ -0,0 +1,2187 @@
+                        let found_leaf =
+                            if node.is_leaf() {
+                                if !cursor.eop() {
+                                    // at an existing leaf with a different path.
+                                    // we're done.
+                                    trace!("Existing but different leaf encountered at {:?} at {:?} -- we're done", &node_ptr, storage.get_cur_block());
+                                    return Err(Error::NotFoundError);
+                                }
+                                else {
+                                    // we're done -- we found the leaf
+                                    true
+                                }
+                            }
+                            else {
+                                false
+                            };
+
+                        if found_leaf {
+                            return Ok((cursor, node));
+                        }
+
+                        // cursor grabbed a copy of node, but not yet a ptr.
+                        // at intermediate node whose child is not present in this trie.
+                        // try to shunt to the prior node that has the child itself.
+                        let (next_node, _, next_node_ptr, _) = MARF::walk_backptr(storage, &node, chr, &mut cursor)?;
+                       
+                        // finish taking the step
+                        cursor.walk_backptr_finish(&next_node_ptr, &storage.get_cur_block());
+
+                        // keep going
+                        node = next_node;
+                        node_ptr = next_node_ptr;
+                    }
+                }
+            }
+        }
+        
+        trace!("Trie has a cycle");
+        return Err(Error::CorruptionError("Trie has a cycle".to_string()));
+    }
+
+    pub fn format(storage: &mut TrieFileStorage, first_block_hash: &BlockHeaderHash) -> Result<(), Error> {
+        storage.format()?;
+        storage.extend_to_block(first_block_hash)?;
+        let node = TrieNodeType::Node256(TrieNode256::new(&vec![]));
+        let hash = get_nodetype_hash(&node, &vec![]);
+        let root_ptr = storage.root_ptr();
+        storage.write_nodetype(root_ptr, &node, hash)
+    }
+
+    pub fn get_path(storage: &mut TrieFileStorage, block_hash: &BlockHeaderHash, path: &TriePath) -> Result<Option<TrieLeaf>, Error> {
+        trace!("MARF::get_path({:?}) {:?}", block_hash, path);
+
+        storage.open_block(block_hash, false)?;
+        let (cursor, node) = MARF::walk(storage, block_hash, path)?;
+
+        if cursor.block_hashes.len() + 1 != cursor.node_ptrs.len() {
+            trace!("cursor.block_hashes = {:?}", &cursor.block_hashes);
+            trace!("cursor.node_ptrs = {:?}", cursor.node_ptrs);
+            assert!(false);
+        }
+
+        if cursor.eop() {
```

Comment:
Will `MARF::walk` ever return `Ok` if `cursor.eop()` isn't true?

Reply:
No -- if it returns `Ok`, then `cursor.eop()` _must_ be true.  Will replace with an `assert!(cursor.eop())`.

Reply:
Should be fixed now.